ffmpeg_new.c

```diff
@@ -0,0 +1,555 @@
+	int ret;
+
+	int64_t pts = av_rescale_q(offset * AV_TIME_BASE, AV_TIME_BASE_Q, st->time_base);
+
+	avcodec_flush_buffers(priv->codec_context);
+	/* Force reading a new packet in next ffmpeg_fill_buffer(). */
+	priv->input->curr_pkt_size = 0;
+
+	ret = av_seek_frame(priv->input_context, priv->input->stream_index, pts, 0);
+
+	if (ret < 0) {
+		return -IP_ERROR_FUNCTION_NOT_SUPPORTED;
+	} else {
+		ffmpeg_buffer_flush(priv->output);
+		return 0;
+	}
+}
+
+static void ffmpeg_read_metadata(struct growing_keyvals *c, AVDictionary *metadata)
+{
+	AVDictionaryEntry *tag = NULL;
+
+	while ((tag = av_dict_get(metadata, "", tag, AV_DICT_IGNORE_SUFFIX))) {
+		if (tag->value[0])
+			comments_add_const(c, tag->key, tag->value);
+	}
+}
+
+static int ffmpeg_read_comments(struct input_plugin_data *ip_data,
+		struct keyval **comments)
+{
+	struct ffmpeg_private *priv = ip_data->private;
+	AVFormatContext *ic = priv->input_context;
+
+	GROWING_KEYVALS(c);
+
+	ffmpeg_read_metadata(&c, ic->metadata);
+	for (unsigned i = 0; i < ic->nb_streams; i++) {
+		ffmpeg_read_metadata(&c, ic->streams[i]->metadata);
+	}
+
+	keyvals_terminate(&c);
+	*comments = c.keyvals;
+
+	return 0;
+}
+
+static int ffmpeg_duration(struct input_plugin_data *ip_data)
+{
+	struct ffmpeg_private *priv = ip_data->private;
+	return priv->input_context->duration / AV_TIME_BASE;
+}
+
+static long ffmpeg_bitrate(struct input_plugin_data *ip_data)
+{
+	struct ffmpeg_private *priv = ip_data->private;
+	long bitrate = priv->input_context->bit_rate;
+	return bitrate ? bitrate : -IP_ERROR_FUNCTION_NOT_SUPPORTED;
+}
+
+static long ffmpeg_current_bitrate(struct input_plugin_data *ip_data)
+{
+	struct ffmpeg_private *priv = ip_data->private;
+	AVStream *st = priv->input_context->streams[priv->input->stream_index];
+	long bitrate = -1;
+	/* ape codec returns silly numbers */
+	if (priv->codec->id == FFMPEG_CODEC_ID_APE)
+		return -1;
+	if (priv->input->curr_duration > 0) {
+		double seconds = priv->input->curr_duration * av_q2d(st->time_base);
+		bitrate = (8 * priv->input->curr_size) / seconds;
+		priv->input->curr_size = 0;
+		priv->input->curr_duration = 0;
+	}
+	return bitrate;
+}
+
+static char *ffmpeg_codec(struct input_plugin_data *ip_data)
+{
+	struct ffmpeg_private *priv = ip_data->private;
+	return xstrdup(priv->codec->name);
+}
+
+static char *ffmpeg_codec_profile(struct input_plugin_data *ip_data)
+{
+	struct ffmpeg_private *priv = ip_data->private;
+	const char *profile;
+
+	profile = av_get_profile_name(priv->codec, priv->codec_context->profile);
+
+	return profile ? xstrdup(profile) : NULL;
+}
+
+const struct input_plugin_ops ip_ops = {
+	.open = ffmpeg_open,
+	.close = ffmpeg_close,
+	.read = ffmpeg_read,
+	.seek = ffmpeg_seek,
+	.read_comments = ffmpeg_read_comments,
+	.duration = ffmpeg_duration,
+	.bitrate = ffmpeg_bitrate,
+	.bitrate_current = ffmpeg_current_bitrate,
+	.codec = ffmpeg_codec,
+	.codec_profile = ffmpeg_codec_profile
+};
+
+const int ip_priority = 300;
```

Comment:
Whoa! I'd suggest we keep the old priority, until we will be *very* confident that the plugin does indeed improve the status quo.

Also, changing the priority should probably cause a cache version bump, to force metadata re-reading.